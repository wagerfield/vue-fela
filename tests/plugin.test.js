import { createLocalVue } from 'vue-test-utils'
import { createRenderer } from 'fela'
import VueFela from '../index'
import {
  FelaRenderer,
  installPlugin,
  wrapComponent,
  createComponent
} from './test-helpers'

describe('Plugin', () => {

  it('calls plugin install method', () => {
    const localVue = createLocalVue()
    const spy = jest.spyOn(VueFela, 'install')
    expect(spy).toHaveBeenCalledTimes(0)
    localVue.use(VueFela)
    expect(spy).toHaveBeenCalledTimes(1)
    spy.mockRestore()
  })

  it('throws error when install is called more than once', () => {
    const localVue = createLocalVue()
    const spy = jest.spyOn(VueFela, 'install')
    VueFela.install(localVue)
    expect(() => {
      VueFela.install(localVue)
    }).toThrow()
    expect(spy).toHaveBeenCalledTimes(2)
    spy.mockRestore()
  })

  it('does not set $fela property on component instances', () => {
    const localVue = installPlugin()
    const component = createComponent('div')
    const wrapper = wrapComponent(component, localVue)
    expect(wrapper.vm.$fela).toBeUndefined()
  })

  it('sets $fela property on component instances', () => {
    const localVue = installPlugin()
    const fela = createRenderer()
    const inner = createComponent('div')
    const outer = createComponent('div', null, [ inner ])

    const wrapper = wrapComponent(outer, localVue, fela)
    expect(wrapper.vm.$options.fela).toBe(fela)
    expect(wrapper.vm.$fela).toEqual(FelaRenderer)
    expect(wrapper.vm.$fela).toBe(fela)

    const child = wrapper.find(inner)
    expect(child.vm.$options.parent.$fela).toBe(fela)
    expect(child.vm.$fela).toEqual(FelaRenderer)
    expect(child.vm.$fela).toBe(fela)
  })

  it('provides fela renderer to children via inject', () => {
    const fela = createRenderer()
    const localVue = installPlugin()

    const child1 = createComponent('div')
    const child2 = createComponent('div', null, null, { inject: [ 'fela' ] })
    const parent = createComponent('div', null, [ child1, child2 ])

    const parentWrapper = wrapComponent(parent, localVue, fela)
    const childWrapper1 = parentWrapper.find(child1)
    const childWrapper2 = parentWrapper.find(child2)

    expect(parentWrapper.vm.fela).toBeUndefined()
    expect(childWrapper1.vm.fela).toBeUndefined()
    expect(childWrapper2.vm.fela).toBe(fela)
  })

  it('registers a fela component in the global scope', () => {
    const localVue = installPlugin()
    expect(localVue.options.components.fela).toBeDefined()
  })
})
