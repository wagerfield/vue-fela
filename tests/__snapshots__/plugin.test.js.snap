// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Plugin server adds head meta function to all component instances 1`] = `
Object {
  "style": Array [
    Object {
      "cssText": ".a{width:8px}.b{color:red}",
      "data-fela-rehydration": 2,
      "data-fela-type": "RULE",
      "hid": "rule",
      "type": "text/css",
    },
  ],
}
`;

exports[`Plugin server adds head meta function to all component instances 2`] = `Object {}`;

exports[`Plugin server adds head meta function to all component instances 3`] = `Object {}`;

exports[`Plugin server uses metaTagId and metaKeyName options 1`] = `
Object {
  "style": Array [
    Object {
      "cssText": ".a{width:8px}.b{color:red}",
      "data-fela-rehydration": 2,
      "data-fela-type": "RULE",
      "type": "text/css",
      "vmid": "rule",
    },
  ],
}
`;

exports[`Plugin throws error if a renderer is not passed as an option 1`] = `"[vue-fela] requires a renderer option"`;

exports[`Plugin throws error when install is called more than once 1`] = `"[vue-fela] already installed. Vue.use(VueFela) should only be called once."`;
